### ISSUE TITLE
Typha in new cluster fails to initialize

### ISSUE THREAD

**Issue by davidmccormick** [projectcalico/typha#118]
## Expected Behavior
Launch new clusters with Typha from empty backend store (I'm using kubernetes as the store)

## Current Behavior
When starting I'm getting the folllowing error in the logs..

```
2018-03-02 11:00:39.028 [INFO][7] daemon.go 225: Using Kubernetes API datastore, checking if we need to migrate v1 -> v3
2018-03-02 11:00:39.056 [WARNING][7] migrate.go 723: Unable to parse Calico version :  is not in dotted-tri format
2018-03-02 11:00:39.057 [ERROR][7] migrate.go 665: Could not parse CalicoVersion  in ClusterInformation: Error converting version
2018-03-02 11:00:39.057 [ERROR][7] daemon.go 251: Failed to determine migration requirements error=Error converting version
2018-03-02 11:00:40.060 [WARNING][7] migrate.go 723: Unable to parse Calico version :  is not in dotted-tri format
2018-03-02 11:00:40.060 [ERROR][7] migrate.go 665: Could not parse CalicoVersion  in ClusterInformation: Error converting version
2018-03-02 11:00:40.060 [ERROR][7] daemon.go 251: Failed to determine migration requirements error=Error converting version
```

## Possible Solution
What happens if Typha is the first thing to connect, should/could it define the cluster?  I'm presuming that the crd.projectcalico.org/v1 ClusterInformation is missing?

## Steps to Reproduce (for bugs)

1.  Spinning up cluster with kube-aws and custom networking daemonsets.
2. Enabling Typha in calico-node with kubernetes backend
3.  Creating Typha deployment.

## Context
I'm trying to create a PR for kube-aws that replaces the external flannel and calico implementation with a canal model with k8s as the datastore.

## Your Environment
* Typha version (if using Typha binary directly `calico-typha --version`): 

Using these versions of the tools...
```
			CalicoNodeImage: model.Image{Repo: "quay.io/calico/node", Tag: "v3.0.3", RktPullDocker: false},
			CalicoCniImage:  model.Image{Repo: "quay.io/calico/cni", Tag: "v2.0.1", RktPullDocker: false},
			FlannelImage:    model.Image{Repo: "quay.io/coreos/flannel", Tag: "v0.9.1", RktPullDocker: false},
			FlannelCniImage: model.Image{Repo: "quay.io/coreos/flannel-cni", Tag: "v0.3.0", RktPullDocker: false},
			TyphaImage:      model.Image{Repo: "quay.io/calico/typha", Tag: "v0.6.2", RktPullDocker: false},
```

* Which orchestrator are you using (e.g. Kubernetes, OpenStack, Docker, Mesos): Kubernetes v1.8.7 (kube-aws)
* Orchestrator version (if used): 
* Etcd version (if using etcd datastore driver): na
* Operating System and version: coreos stable
* Link to your project (optional): https://github.com/kubernetes-incubator/kube-aws

**Comment by davidmccormick** [comment]
I've noticed that when the cluster is stuck in this non-working state that the clusterinformation 'default' object is missing it's calicoVersion attribute: -

```
kubectl get clusterinformations default -o yaml
apiVersion: crd.projectcalico.org/v1
kind: ClusterInformation
metadata:
  annotations:
    projectcalico.org/metadata: '{"uid":"2ba5da47-2089-11e8-bcf4-066a14973e7e","creationTimestamp":"2018-03-05T15:23:34Z"}'
  clusterName: ""
  creationTimestamp: 2018-03-05T15:23:34Z
  deletionGracePeriodSeconds: null
  deletionTimestamp: null
  initializers: null
  name: default
  namespace: ""
  resourceVersion: "4568"
  selfLink: /apis/crd.projectcalico.org/v1/default
  uid: 2b2e4386-2089-11e8-954b-0a5f3cd76b7e
spec:
  clusterGUID: 20d6c133f33d408784af0125368b1c9f
  clusterType: typha,kdd
  datastoreReady: true
```
This might account for the success at reaching the datastore (apiserver) and yet the failure in working out the version number.  How does this record get population when starting up a new cluster?  Could it be a cluster coming up with Typha is managing to write this missing the calicoVersion string?

**Comment by fasaxc** [comment]
@davidmccormick Thanks for the report.  Just so we're in sync, what was the previous state of this cluster?  Was it a v2.6 cluster or is it a new cluster?

Typha doesn't know the Calico version because the Calico version is only defined in calico/node.  That's why it doesn't fill that in (but calico/node should be dealing with that case).

**Comment by davidmccormick** [comment]
It was a v2.6.5 cluster but with an etcd datastore so the clusterinformation is being written for the first time.  I have a work-around where I force the masters to not use Typha and make sure that they are started before Typha gets a chance to write the bad record.

Maybe Typha should connect and set the clusterinformation default crd until a calico client connects to it and tells it what version number it is?  The issue is definitely that Typha is generating the default clusterinformation with a missing CalicoVersion attribute.

**Comment by fasaxc** [comment]
I don't think we support migration from one store to another, i think we expect you to go from etcdv2 to etcdv3 or k8s to k8s.  @caseydavenport can you confirm?

**Comment by davidmccormick** [comment]
@fasaxc I would still have the same problem if I wanted to spin a cluster up from scratch with Typha - forget the fact that I'm upgrading.  Whenever Typha starts up before a calico has written a clusterinformation record it will create a bad one.

Why force people to start a calico client before typha - that's procedural rather than declarative, so not so much the kubernetes way! :)

**Comment by caseydavenport** [comment]
> I don't think we support migration from one store to another, i think we expect you to go from etcdv2 to etcdv3 or k8s to k8s. @caseydavenport can you confirm?

That's correct, migration from k8s -> etcd and vice-versa isn't supported. Not sure how one would make that work, but if you do happen to get it working I'd love to know how :)

> Why force people to start a calico client before typha - that's procedural rather than declarative, so not so much in the spirit of the kubernetes way! :)

What I'd expect to happen is that Typha would exit because it can't find the configuration (as is described in this issue) and Typha would be restarted until calico/node writes the clusterinfo, at which point Typha would come up just fine. If that's not happening then we probably need to investigate why.

**Comment by davidmccormick** [comment]
Adding an init container to delay the startup seems to work as a work-around for now...

```
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: calico-typha
  namespace: kube-system
  labels:
    k8s-app: calico-typha
spec:
...
      initContainers:
      - name: wait-for-default-clusterinformation
        image: quay.io/coreos/hyperkube:v1.8.7_coreos.0
        command: ["/bin/sh", "-c" ]
        args:
          - |
              set -x ;
              while true; do
                if /kubectl get clusterinformation default; then
                  echo "Calcio clusterinformation default exists, can continue..."
                  exit 0
                fi
                echo "Waiting for Calcio clusterinformation default..."
                sleep 5
              done
```

**Comment by caseydavenport** [comment]
@davidmccormick theoretically we could add something like that into the Typha code itself, but I'm a bit surprised that Typha wouldn't recover once calico/node starts up.

Are you seeing Typha continue to crash even after calico/node starts?

**Comment by davidmccormick** [comment]
@caseydavenport yes.  Once Typha has written the bad clusterinformation default object then neither calico node or typha will then start up.  They both cycle with crashloop back-off with the "Unable to parse Calico version :  is not in dotted-tri format" error.

If Typha has to wait for a calico-client to start up and set the clusterinformation record - that means you are going to have to have two different calico node deployments in a cluster one connecting directly and the other via Typha ... unless there is a way to get felix to start in an init container to set the record and then exit?

**Comment by caseydavenport** [comment]
The ClusterInformation should be written before Felix even starts, because it's done as part of calico/node startup: https://github.com/projectcalico/calico/blob/master/calico_node/startup/startup.go#L924-L927

So I'm not seeing the cyclical dependency... I'll try to find some time to reproduce to see if I can spot what's going on.

**Comment by davidmccormick** [comment]
> That's correct, migration from k8s -> etcd and vice-versa isn't supported. Not sure how one would make that work, but if you do happen to get it working I'd love to know how :)

Hi @caseydavenport for information - I got the migration from flannel+calico with etcd backend to kubernetes daemonset with kube-api as backend.  Admittedly this is in clusters without ANY custom  calicoctl config and few NetworkPolicy rules.

In order to migrate, I found it necessary to run a script which examines each cluster node for running pods (or run a new pod) in order to deduce which podCIDR flannel has allocated (I didn't fancy attaching to etcd).   For each node, I patch the spec.podCIDR so that kubernetes knows this when we migrate the networking.  Then rolling out the deamonsets are then straight-forward - deleting the existing cni config and calico daemonset and replacing with either canal or plain flannel. Typha is optional (and uses the init container above to stop it corrupting the cluster version).  Masters are configured to not use Typha so that they come up and set cluster version (needed because of this issue).

So far testing migrations have resulted in very little cluster disruption. :) Roll-backs are a lot more disruptive and require that all nodes get rolled.

**Comment by caseydavenport** [comment]
fyi I believe this should be fixed by this PR: https://github.com/projectcalico/libcalico-go/pull/841

It's been backported to the release-v3.0 branch so it should be fixed in v3.0.5, which I'm expecting to cut within the next week.

**Comment by caseydavenport** [comment]
v3.0.5 has been released, so I'm going to close this. Please report back if you still see this!
